enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,483 @@
+|None platform SNO only day2 workers     |None    |0  |1+ |SR |SR |W  |M  |Yes, ingress floats                              |No, ingress pinned to master pool             |Ingress issue, unsupported, untested|
+|Cloud platform SNO only day2 workers    |Cloud   |0  |1+ |SR |SR |W  |M  |Yes - provided, ingress floats                   |Yes - provided, ingress floats                |Works, unsupported, untested        |
+
+Some notes about this table and the decisions made in it:
+
+The table assumes that the "Infrastrcture Topology" value is read only. That
+means an optimal value for this parameter must be determined during
+installation. You can see the consequences of that, for example, in the last
+row of the table - even though we add workers in day 2 to the cluster, the
+"Infrastrcture Topology" stays "SingleReplica" even though that cluster could
+benefit from the additional worker nodes for the purposes of highly-available
+infrastructure (as it's running in the cloud, so it has a load-balancer). In
+the future it may be worth revisiting making this value configurable.
+
+The "None platform SNO day1 single/multiple worker" scenario rows aren't really
+possible today as there is no real way to install this kind of cluster. That's
+why their CIT / CTP have been left empty.
+
+It's proposed that the only scenarios in which the "Infrastrcture Topology"
+parameter will be "HighlyAvailable" are when installing a single control-plane
+node cluster in the cloud with one or more day-1 workers. Cloud deployments
+have a load balancer so they can benefit from having highly-available ingress.
+
+Today, if you only have 1 day-1 worker in a single control-plane node cluster
+in the cloud, then the "Infrastrcture Topology" is set to "SingleReplica" - I
+don't believe there's a good reason for that, as both the control-plane node
+and the worker node are actually in the worker pool (TODO: make sure this is
+correct!, if not, this entire point is moot) - giving us a total of 2 worker
+nodes, so there's nothing preventing us from setting "Infrastrcture Topology"
+to "HighlyAvailable" in that case.
+
+When installing day-1 workers on a single control-plane node cluster on the
+"none"-platform, it's unlikely you have a load balancer / virtual IP set up, so
+even if you have multiple workers to hold ingress pods, there would be no load
+balancer to balance between them. So it makes sense to only have one ingress
+replica. That's why we set "Infrastrcture Topology" to "SingleReplica" even in
+the "none"-platform even though we have day-1 workers. This has the unfortunate
+consequence that non-ingress infrastructure workloads (such as monitoring, for
+example) can no longer benefit from spreading over the worker nodes - since
+they will all also have just 1 replica (as a result of the topology). It's
```

Comment:
This consequence is a smell. It is what tells me that forcing the infrastructure topology to single replica for the purposes of ensuring that the router pods run on the masters is not a desirable solution. It seems to me that one and only goal of this entire proposal is getting the router pods pinned to run on the master in SNO. The discussion about topologies is just distracting from that goal. Since the real goal is pinning the router pods, and the router pods are strictly under the purview of the ingress operator, the configuration directing the ingress operator to place the router pods on the master should be in the ingress.config.openshift.io resource. The configuration of the router placement then has no side-effects for the rest of the cluster.

Reply:
For the record Ben shared similar concerns to which I responded [here](https://github.com/openshift/enhancements/pull/996#discussion_r787631818) (third quote) 

The thought process behind changing the topology was -

- "We need to pin the router pod to master in "none" platform because SNO users typically don't want / need a load balancer / VIP solution by default even when there are workers"

- "It's insufficient to simply pin the router to master because if there are multiple workers in day 1 the installer would set infra topology to HA and then we would have multiple replicas of the router pod which doesn't make sense if they're all pinned to a single node"

- "We need to avoid having multiple router replicas. The only solution: Set infra topology to SingleReplica to avoid multiple routers"

I didn't realize adding new APIs / configuration options would be desirable, so I didn't really consider it as an option at all, so I went with the suboptimal (and only solution I can think of) of "abusing" the infrastructure topology parameter. 

But given your & Ben's response to the proposed solution I guess adding new configuration options is a reasonable thing to do.

_______

I guess what we would want then would be a new way to tell the Ingress Operator (ideally through the IngressController API) to ignore the infrastructure topology altogether when determining the number of replicas of the router deployment and always set the "replicas" parameter to 1 and pin that router pod to master.

This way the installer logic where the number of day 1 workers determines the infra topology stays the same, infrastructure workloads can still spread over all nodes in the worker pool, and ingress would be just 1 replica pinned to master.

This new IngressController configuration, for now, will be applied by the Installer automatically in single control-plane node "none"-platform clusters. In the future if one day the load balancer is removed from cloud IPI deployments of single control-plane node clusters then this configuration could be applied on those platforms as well.

Does this approach make sense?

Reply:
Come to think of it it doesn't necessarily have to be a new configuration option... Maybe simply setting the IngressContrller replicas to 1 and pinning it to master by the installer is all we have to do

Reply:
> I guess what we would want then would be a new way to tell the Ingress Operator (ideally through the IngressController API) to ignore the infrastructure topology altogether when determining the number of replicas of the router deployment and always set the "replicas" parameter to 1 and pin that router pod to master.

The Ingress Operator would not always set the replicas to 1. The Ingress Operator would consider the control plane topology rather than the infrastructure topology when configured to place its pods on the masters.

Reply:
> Maybe simply having the installer set the IngressContrller replicas to 1 and pin it to master is all we have to do?

Directing the ingress operator to "pin it to master" is what requires an api change, unless I am missing something.

Reply:
>The Ingress Operator would not always set the replicas to 1. The Ingress Operator would consider the control plane topology rather than the infrastructure topology when configured to place its pods on the masters.

Yeah that makes more sense

>Directing the ingress operator to "pin it to master" is what requires an api change, unless I am missing something.

We could simply do something like this:
![image](https://user-images.githubusercontent.com/10882062/150657009-63bf23e2-57aa-44d0-adf3-b0bb0e8a2fe5.png)


(don't mind the bip case condition)

Reply:
> We could simply do something like this

That is the action that the ingress operator would take to pin the routes to the masters. But that is not how the ingress operator would know to take that action. That last bit is what would require a new configuration item.

Reply:
@Miciah

Reply:
I do wonder though whether we dont want to have all the infra components pinned to the master - for predictability, ease of debug and keeping the workers for users workloads

Reply:
>That is the action that the ingress operator would take to pin the routers to the masters. But that is not how the ingress operator would know to take that action. That last bit is what would require a new configuration item.

What I'm saying is that the ingress operator doesn't really have to know to take that action, we can simply create this IngressController resource in the installer (the screenshot above is installer code)

Reply:
Ah, sorry, I missed that that was the installer code for generating the default ingress controller. The installer generates a manifest for a default ingress controller only if the cluster is using an internal publishing strategy. The publishing strategy can only be set to internal when installing to a cloud platform. Perhaps with this enhancement, the installer would allow an internal publishing strategy for non-cloud platforms. Nevertheless, I still think that it is better for the ingress operator itself to determine the node placement based on configuration that the ingress operator knows. This would allows us to keep consistent behavior regardless of the publishing strategy.

Reply:
I'm not sure I understand how the publishing strategy is relevant: I know the installer currently only generates an IngressController manifest if the publishing strategy is internal, but how does that relate to the discussion? What stops us from also generating an IngressController manifest when the platform is "none" (regardless of publishing strategy)?

Reply:
There is nothing that stops the installer from doing that. Is that what we want? Does the network edge team have an opinion on whether they would prefer the installer an IngressController or whether they would prefer the ingress operator create the IngressController? My preference is to have the ingress operator do it since it keeps the domain knowledge located in the operator instead of the installer.

> generating an IngressController manifest when the platform is "none" and the control plane topology is "SingleReplica"

Would this be when the platform is an on-prem platform rather than just a "none" platform?

Reply:
> There is nothing that stops the installer from doing that. Is that what we want? Does the network edge team have an opinion on whether they would prefer the installer an IngressController or whether they would prefer the ingress operator create the IngressController? My preference is to have the ingress operator do it since it keeps the domain knowledge located in the operator instead of the installer.
> 
> > generating an IngressController manifest when the platform is "none" and the control plane topology is "SingleReplica"
> 
> Would this be when the platform is an on-prem platform rather than just a "none" platform?

Okay I've gave it some thought, do you personally see any problem with the following approach then:

In the absence of IngressController node-selector and replicas, if the user is on an on-prem platform (none, vsphere, etc*) IngressOperator will "put ingress on the contrl-plane" meaning:
- It will set number of replicas to equal controlplanetopology (1 if singlereplicas 2 if ha)
- It will set nodeSelector to master rather than worker

No installer changes, no new ingresscontroller API (other than changing documentation in the existing API to reflect that behavior)

* I guess IngressController would also have to maintain a list of platforms it considers "on-prem" to do that

Reply:
> Does the network edge team have an opinion on whether they would prefer the installer an IngressController or whether they would prefer the ingress operator create the IngressController?

My opinion is that the proposed defaulting semantics for the IngressController API are somewhat complex (based on the table under "Infrastructure Topology value"), and this could be rather difficult to explain in the API godoc and difficult for the cluster administrator to understand.  

It seems to me that instead of using some non-obvious inference rules to define a function to map a cluster infrastructure config to a set of nodes on which router pods should run, it would be more obvious and more in the Kubernetes way to label the nodes at install-time (using `node-role.kubernetes.io/infra`, or some novel label if the definitions of "infra node" and "router node" should be independent) and allow the user to adjust labels as desired after installation.  Would using labels in this way be out of the question?

Reply:
> *I guess IngressController would have to maintain a list of platforms it considers "on-prem" to do that

We kind of do that already to decide whether to default the endpoint publishing strategy type to `HostNetwork` or to `LoadBalancerService`.  

To my point in my previous comment, the current defaulting logic is straightforward enough (platform type → endpoint publishing strategy type, infrastructure topology → number of replicas).  The propose logic will be more complex ((platform type, control-plane topology, infrastructure topology) → node selector).  Complex defaulting logic can lead to surprises.

Reply:
>My opinion is that the proposed defaulting semantics for the IngressController API are somewhat complex (based on the table under "Infrastructure Topology value"), and this could be rather difficult to explain in the API godoc and difficult for the cluster administrator to understand.

A lot of that complexity you're describing is coming from me trying to abuse the Infrastructure Topology value to achieve my goals, @staebler's proposal is to admit that infra topology has nothing to do with it - and just tell Ingress Operator to completely disregard infrastructure topology when the platform is not a cloud platform. Instead it should look at the platform and control-plane topology, because when the control-plane topology is "SingleReplica" and the platform is on-prem, the ingress is practically bound to the control-plane due to the likely absence of a load-balancer/VIP, so it shouldn't care about the infra topology.

The infra topology is still important for other things like monitoring, logging, etc - they can still regard it, their behavior is not bound to an existence of a load-balancer

Reply:
Edited my previous [comment](https://github.com/openshift/enhancements/pull/996#discussion_r799388937) since it was a little wrong

Reply:
To summarize the ingress logic, it would look something like this:
- If on-prem platform and single-replica control plane topology, set replicas to 1 and pin to master
- If cloud platform, set replicas according to infrastructure topology and pin to workers

Does that still sound too complicated?

Reply:
As an exercise, I tried to write out my understanding of the proposal for `spec.replicas` as I would expect it to be documented in the API:

> If unset, the operator uses a default value that depends on the platform type, control-plane topology, and infrastructure topology.
> 
> For cloud platforms, the default value depends on the infrastructure topology on the assumption that the cluster has dynamically configured DNS and a cloud load-balancer that can dynamically add worker nodes to its pool.  For on-premise platform, the default value depends on the control-plane topology on the assumption that a single-node cluster may have static DNS and load-balancer configuration (or no load balancer at all) with the ingress controller pinned to the control-plane node, and so adding additional workers should not change the number of ingress controller replicas.
> 
> The following platform types are considered to be cloud platforms: Alibaba, AWS, Azure, GCP, IBMCloud, and PowerVS.  All other platform types (unspecified, BareMetal, OpenStack, etc.) are considered to be on-premise platforms.
> 
> If the platform type is a cloud platform and the infrastructure topology is "SingleNode", then the default number of replicas is 1.  If the platform type is on-premise and the control-plane topology is "SingleNode", then the default number of replicas is 1. Otherwise, the default number of replicas is 2.

We would want similar godoc for the `spec.nodeSelector` field.  

Does my example capture the full complexity?  Do you anticipate any future changes that could make the defaulting logic still more complex?  

I'm still a little reluctant to put this much complexity behind API field defaulting, but if it passes API review, we can include it in the operator.

Reply:
I do not think that the ingress operator should be inferring from platform type where to put the router pods. My suggestion is that we add a field to the configuration of the ingress operator to direct the pods to either the workers or the masters.
The logic would be the following.
* If the ingress operator is configured for routing on masters, the ingress operator will use a node selector for the pods to place them on the masters and will set the replica count based on the ControlPlaneTopology.
* If the ingress operator is configured for routing on workers, the ingress operator will use a node selector for the pods to place them on the workers and will set the replica count based on the InfrastructureTopology. My assumption there is that is the current behavior. Please correct my naive understanding of what the ingress operator does under the covers.
* For capability with upgrades, if there is no configuration set, then the infgess operator will default to workers for the configuration.

Reply:
> add a field to the configuration of the ingress operator to direct the pods to either the workers or the masters.

I see. I guess I just find that idea a bit strange because, in a way, the IngressController already has an API to do this - it's called the nodeSelector.

You suggest we would introduce another, more explicit "master vs workers" field, that will control the defaulting behavior of both the replicas and the nodeSelector. Sounds reasonable.

Here's another idea I want to float, it's a bit hacky, let me know what you think -
Instead of introducing a new  API parameter, the ingress operator would adjust its defaulting `replicas` behavior depending on whether the IngressController nodeSelector has particular labels or not (master/worker labels).

It's a bit ugly because as a user I would find it surprising that my choice of node selector labels is somehow able to affect other fields of the IngressController, but on the other hand it would get rid of a bit of redundancy in the IngressController API

Reply:
> I see. I guess I just find that idea a bit strange because, in a way, the IngressController already has an API to do this - it's called the nodeSelector.

Yes, the IngressController has an API to do this. And I would expect the ingress operator to use that. I prefer that the ingress operator create the IngressController rather than having the installer create it. As far as I know, ingress is the only operator for which the installer is creating controllers. I don't know the history of that, but unless there is strong desire by the ingress team to double down on that, then I would rather not add to the criteria in which the installer needs to create an IngressController.

Reply:
> Instead of introducing this new boolean parameter to IngressController

I am not suggesting adding a field to IngressController. I am suggesting adding a field to the [Ingress CRD](https://github.com/openshift/api/blob/master/config/v1/types_ingress.go#L16).

Reply:
Oh, didn't realize that you're talking about the Ingress CRD. sgtm.

Reply:
>Add a high-level "ingress topology" field analogous to the control-plane topology and infrastructure topology fields; the ingress operator could use this high-level description of intent in its defaulting logic.
Add both a nodePlacement field and a replicas field to the ingress config; the ingress operator could directly use these values, if provided, as default values for the corresponding fields in ingresscontrollers.

I'm leaning towards the first option, since the second one would put the burden of understanding the complexities involved on the installer, because it would be the one who would have to set the correct values on that ingress config CR.

With the first option all the installer has to do is set the "IngressTopology" to "ControlPlane"

>One question: Would these values only be used when generating a default ingresscontroller, or would they be used when defaulting the spec.replicas and spec.nodePlacement fields on any ingresscontroller that doesn't specify these fields?

I think the answer that makes most sense is "both", is there any concerns that you have with that? I can't come up with any personally

Reply:
>Add a high-level "ingress topology" field analogous to the control-plane topology and infrastructure topology fields; the ingress operator could use this high-level description of intent in its defaulting logic.

Ran this by @dhellmann , he proposed to have this topology field under the Ingress CRD rather than be a "high level" API as you mentioned, since it's component-specific.

I think that makes sense, doing that would be kind of a hybrid between your two proposed solutions - we make it abstract (a boolean topology) - but we put it in the Ingress config CR rather than in a high level cluster configuration.

The installer can then not care about the exact defaulting details (exact label values and number of replicas)

Reply:
> I think the answer that makes most sense is "both", are there any concerns that you have with that? I can't come up with any personally

I have no concerns with this answer; this approach makes sense.  

> Ran this by @dhellmann , he proposed to have this topology field under the Ingress CRD rather than have it be a "high level" API as you mentioned, since it's component-specific.
> 
> I think that makes sense, doing that would be kind of a hybrid between your two proposed solutions - we make it abstract (a boolean topology) - but we put it in the Ingress config CR rather than in a high level cluster configuration.

This is actually what I meant.  Sorry that I wasn't clearer.  We are in complete agreement.